File: python_version/arrays/3 SUM.py
class Solution:
    """
    @param numbers: Give an array numbers of n integer
    @return: Find all unique triplets in the array which gives the sum of zero.
    """

    def threeSum(self, numbers):
        # write your code here
        # combination of double pointer and hashmap
        numbers.sort()
        result = []
        for i in range(len(numbers) - 2):
            left, right = i + 1, len(numbers) - 1
            while left < right:
                if numbers[left] + numbers[right] + numbers[i] > 0:
                    right -= 1

                elif numbers[left] + numbers[right] + numbers[i] < 0:
                    left += 1

                elif numbers[left] + numbers[right] + numbers[i] == 0:
                    if [numbers[i], numbers[left], numbers[right]] in result:
                        pass
                        left += 1
                        right -= 1
                    else:
                        result.append([numbers[i], numbers[left], numbers[right]])
                        left += 1
                        right -= 1

        return result

    def threeSumClosest(self, numbers, target):
        # write your code here
        numbers.sort()
        result = None
        for i in range(len(numbers) - 2):
            left, right = i + 1, len(numbers) - 1
            while left < right:
                amount = numbers[i] + numbers[left] + numbers[right]

                if result is None:
                    result = amount

                if abs(result - target) > abs(amount - target):
                    result = amount

                if amount < target:
                    left += 1

                elif amount > target:
                    right -= 1

                elif amount == target:
                    return amount

        return result

File: src/two-pointer/3-sum.java
import java.util.*;

class Solution {
    public List<List<Integer>> threeSum(int[] arr) {
        List<List<Integer>> triplets = new ArrayList<>();
        for(int i = 0; i < arr.length - 2; i++){
          if(i > 0 && arr[i] == arr[i - 1]){
            continue;
          }
          Arrays.sort(arr);
          int left = i + 1, right = arr.length - 1;

          while(left < right){
            int temp = arr[i] + arr[left] + arr[right];
            if(temp == 0){
              triplets.add(Arrays.asList(arr[i], arr[left], arr[right]));
              left += 1;
              right -= 1;
              while(left < right && arr[left] == arr[left - 1]){
                left += 1;
              }
              while(left < right && arr[right] == arr[right + 1]){
                right -= 1;
              }
            }
            else if(temp > 0){
              right -= 1;
              while(left < right && arr[right] == arr[right + 1]){
                right -= 1;
              }
            }
            else{
              left += 1;
              while(left < right && arr[left] == arr[left - 1]){
                left += 1;
              }
            }
          }
        }
        return triplets;
    }
}